PyTorch migration: Add FETA estimators

timokau opened the pull request:
## Description

This adds pytorch versions of the FETA estimators. ~It builds on (and currently includes the commits of) #164, which should be reviewed first.~ This is now ready for review.

## Motivation and Context

Part of the ongoing pytorch migration.

## How Has This Been Tested?

I have added the FETA estimators to the existing tests for discrete choice, general choice and object ranking. I used the test suite and the linters to check the changes.

## Does this close/impact existing issues?

Related to #125.

## Types of changes

This depends on what you consider as the base. It is a new feature for the pytorch migration branch, but a breaking change on master.

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.

kiudee commented:
Looks good to me. Will finalize the review as soon as the branch is rebased on the pytorch branch after the migration/FATE branch has been merged.
I assume the target of the pull request should be kiudee:pytorch-migration, right?

timokau commented:
> Looks good to me. Will finalize the review as soon as the branch is rebased on the pytorch branch after the migration/FATE branch has been merged.

I have rebased this on the merged version of #164. It's only a naive "syntactic" rebase though. There are still some changes to make to adapt FETA to that version. Some polish is needed as well. I will notify you when this PR is ready for review.

> I assume the target of the pull request should be kiudee:pytorch-migration, right?

Yes, thanks for noticing. Changed.

timokau commented:
Rebased, adapted & polished. This is now ready for review.

kiudee commented on csrank/choicefunction/feta_choice.py:
```diff
@@ -0,0 +1,80 @@
+
+    Parameters
+    ----------
+    n_hidden_set_layers : int
+        The number of hidden layers that should be used for the ``DeepSet``
+        context embedding.
+
+    n_hidden_set_untis : int
+        The number of units per hidden layer that should be used for the
+        ``DeepSet`` context embedding.
+
+    n_hidden_joint_layers : int
+        The number of hidden layers that should be used for the utility
+        function that evaluates each object in the aggregated context.
+
+    n_hidden_joint_units : int
+        The number of units per hidden layer that should used for the utility
+        function that evaluates each object in the aggregated context.
```
The docstring is copied from FATE apparently.

timokau replied:
I meant to revisit the docstrings but then forgot to actually do it. I should have caught that in self-review. Thanks for noticing.

timokau replied:
Changed.

kiudee commented on csrank/discretechoice/feta_discrete_choice.py:
```diff
@@ -0,0 +1,78 @@
+
+    Parameters
+    ----------
+    module : torch module (class)
+        This is the scoring module. It should be an uninstantiated
+        ``torch.nn.Module`` class that expects the number of features per
+        object as its only parameter on initialization.
+
+    choice_size : int
+        The size of the target choice set.
+
+    criterion : torch criterion (class)
+        The criterion that is used to evaluate and optimize the module.
+
```
Some parameters are missing here.

timokau replied:
Yes, the documentation was out of date. Thanks for noticing. Changed.

kiudee commented on csrank/modules/collection.py:
```diff
@@ -0,0 +1,134 @@
+
+
+class PairwiseEvaluation(nn.Module):
+    """A module that applies a first-order utility to all object pairs.
+
+    This module receives a list of instances as an input:
+
+    >>> object_00 = [0.5, 0.6]
+    >>> object_01 = [1.5, 1.6]
+    >>> object_02 = [2.5, 2.6]
+    >>> object_10 = [3.5, 3.6]
+    >>> object_11 = [4.5, 4.6]
+    >>> object_12 = [5.5, 5.6]
+    >>> # instance = list of objects to rank
+    >>> instance_a = [object_00, object_01, object_02]
+    >>> instance_b = [object_10, object_11, object_12]
+    >>> instances = torch.tensor([instance_a, instance_b])
+    >>> n_features = instances.size(2)
+
+    And computes a list of *relations*, one relation for each instance. A
+    relation is a 2d array that contains a score for each pairwise object
+    combination.
+
+    >>> import numpy as np
+    >>> from csrank.modules.object_mapping import DeterministicSumming
+    >>> utility = PairwiseEvaluation(DeterministicSumming(2 * n_features))
+    >>> utilities = utility(torch.tensor(instances))
+    >>> utilities
+    tensor([[[ 2.2000,  4.2000,  6.2000],
+             [ 4.2000,  6.2000,  8.2000],
+             [ 6.2000,  8.2000, 10.2000]],
+    <BLANKLINE>
+            [[14.2000, 16.2000, 18.2000],
+             [16.2000, 18.2000, 20.2000],
+             [18.2000, 20.2000, 22.2000]]])
+    >>> utilities[0][0][1] == np.sum(object_00) + np.sum(object_01)
+    tensor(True)
```
As always, great documentation!

timokau replied:
Thanks :)

kiudee reviewed (changes requested):
Except for the small issues in the docstrings, this looks good to go.

timokau commented:
I have addressed your review and also extended the tests to always cover the configurations with and without a zeroth order model (as discussed in chat). I checked that the linters and tests have no complaints again. Please take another look.

timokau commented:
Thank you for the review @kiudee :)

timokau commented:
I missed the documentation of the `activation` estimator parameter. Follow-up in #185.

timokau commented:
I fixed another documentation mistake in #187.
